sdk/include/ds/linked_list.h

```diff
@@ -0,0 +1,642 @@
+		 * the encoded forms are also representing other state (for example,
+		 * bit-packed flags in pointer address bits).
+		 */
+		template<typename T>
+		concept HasPrimOps = requires(T &t)
+		{
+			/** Proxies for list linkages */
+			{
+				t.cell_next()
+				} -> ds::pointer::proxy::Proxies<T>;
+			{
+				t.cell_prev()
+				} -> ds::pointer::proxy::Proxies<T>;
+		};
+
+		/**
+		 * Initialize to singleton ring.  Not all cons cells are required to be
+		 * able to do this, though if you're sticking to rings and not (ab)using
+		 * the machinery here in interesting ways, this should be easy to
+		 * specify.
+		 */
+		template<typename T>
+		concept HasInit = requires(T &t)
+		{
+			{
+				t.cell_init()
+				} -> std::same_as<void>;
+		};
+
+		template<typename T>
+		concept HasPrimOpsInit = HasPrimOps<T> && HasInit<T>;
+
+		/**
+		 * Additional, optional overrides available within implementation of
+		 * cons cells.  It may be useful to static_assert() these in
+		 * implementations to make sure we are not falling back to the defaults
+		 * in terms of the above primops.
+		 *
+		 * @{
+		 */
+		template<typename T>
+		concept HasIsSingleton = requires(T &t)
+		{
+			{
+				t.cell_is_singleton()
+				} -> std::same_as<bool>;
+		};
+
+		template<typename T>
+		concept HasIsSingletonCheck = requires(T &t)
+		{
+			{
+				t.cell_is_singleton_check()
+				} -> std::same_as<bool>;
+		};
+
+		template<typename T>
+		concept HasIsDoubleton = requires(T &t)
+		{
+			{
+				t.cell_is_doubleton()
+				} -> std::same_as<bool>;
+		};
+
+		template<typename T>
+		concept HasEqOps = requires(T &t)
+		{
+			{
+				t.cell_next_is_eq(&t)
+				} -> std::same_as<bool>;
+
+			{
+				t.cell_prev_is_eq(&t)
+				} -> std::same_as<bool>;
+		};
+
+		/** @} */
+
+	} // namespace cell
+
+	/**
+	 * Self-loops indicate either the sentinels of an empty list or,
+	 * less often, singletons without their sentinels; it's up to
+	 * the caller to know which is being tested for, here.
+	 *
+	 * The default implementation decodes and compares one link;
+	 * implementations may have more efficient mechanisms.
+	 *
+	 * @{
+	 */
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<!cell::HasIsSingleton<T>, bool>
+	                is_singleton(T *e)
+	{
+		return e == e->cell_prev();
+	}
+
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<cell::HasIsSingleton<T>, bool>
+	                is_singleton(T *e)
+	{
+		return e->cell_is_singleton();
+	}
+	/** @} */
+
+	/**
+	 * Like is_singleton(), but checks both edges.  Useful only for
+	 * testing invariants.
+	 *
+	 * The default implementation decodes and compares both links.
+	 */
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<!cell::HasIsSingletonCheck<T>, bool>
+	                is_singleton_check(T *e)
+	{
+		return (e == e->cell_next()) && (e == e->cell_prev());
+	}
+
+	template<cell::HasPrimOps T>
+	__always_inline std::enable_if_t<cell::HasIsSingletonCheck<T>, bool>
```

Comment:
Can this not use a requires clause instead of `enable_if`?

Reply:
Still learning C++.  `enable_if`-s removed.